### ISSUE TITLE
NLP examples with variable input len

### ISSUE THREAD

**Issue by escorciav** [quic/qidk#5]
Dear team,

After taking a look at the NLP solutions: [QA](https://github.com/quic/qidk/tree/master/Solutions/NLPSolution1-QuestionAnswering#11-generate-electra-small-model-as-tensorflow-frozen-graph), and [Sentiment analysis](https://github.com/quic/qidk/tree/master/Solutions/NLPSolution2-SentimentAnalysis#21-create-validation_set-directory-and-execute-the-following-scripts),
none of them support varied input length.

**Does QNN support variable input length tensors?**
New versions do digest graphs with opset >= 9, and opset >= 11 does support variable len axis. Thus, it's opt to your tooling :)

Moreover, the inference of recent NLPmodels, e.g., autoregressive Tansformer decoder, often rely on caching key & values. _Could you please provide any guidance to better support such inference pipeline?_

Thanks in advance & kudosto all the team for providing examples of AI models using Qualcomm hardware :)

**Comment by quic-rneti** [comment]
Hello  Victor Escorcia Castillo - 

>> Does QNN support variable input length tensors?

Current QNN release does not support variant input length tensors. 
We are making one End-to-End application with Whisper ASR model. It will be published shortly. 
If you would like to see examples for any other model (or architecture), let us know. We will try our best to make it happen.

**Comment by escorciav** [comment]
It'd be great to have access to all the demos showcased during the Snapdragon summit of 2023 :smile: 

https://www.youtube.com/watch?v=h_vh7_n_OPs

**Comment by ZodiacFRA** [comment]
@quic-rneti Hi :)

>  We are making one End-to-End application with Whisper ASR model. It will be published shortly.

Any ETA on this? 
Have a great day

**Comment by quic-rneti** [comment]
We are working on this. You can expect the tutorials to be available sometime in next months. (2 weeks to 4 weeks is the lead time)

**Comment by ZodiacFRA** [comment]
Hi @quic-rneti thanks for the response :) Could you by any chance share a lowerable whisper script before releasing the fully fledged app? or the dlc file alone?

Have a great day!

**Comment by escorciav** [comment]
@ZodiacFRA I haven't tested whisper, but please tag me (here:Gh or in Twitter) if you do it.

BTW, if they provide TFLite models & you plan to use QNN, just follow the [tutorial](https://docs.qualcomm.com/bundle/publicresource/topics/80-63442-50/tutorial2.html)  :wink: . Preferably, follow the docs of the qnn that you're using. In my case, it's `2.19.0.x` => `opt/qcom/aistack/qnn/2.19.0.240124/docs/QNN/general/tutorial2.html#id1`.

_Disclaimer._ I haven't personally used TFLite as an ML frontend. But, some colleagues have done.

_Personal opinions/judgements: Food for thought_
1. BTW, I've the impression that TFLite & ONNX are way more cross-platform than PyTorch.
2. _Why do I say so?_
[xnnpack](https://github.com/google/XNNPACK) was/is developed/maintained by Google. I'm assuming TFLite path to land on mobile/web/etc. is cleaner & better tested/maintained.
3. _What's xnnpack?_
High-efficiency floating-point neural network inference operators for mobile, server, and Web

**Comment by escorciav** [comment]
@quic-rneti Is there any relevant example from the [Qualcomm AI Hub](https://github.com/quic/qidk/issues/5#issuecomment-1978153000) related to the topic of the issue?

**Comment by ZodiacFRA** [comment]
@escorciav Thanks for the quick response :)

- About the tutorial: It looks like the conversion step still requires fixed length inputs (like all the snpe-[tflite|onnx|pytorch]-to-dlc scripts), this prevents me from converting whisper as one of the self attention matrix dimensions is the number of generated tokens and cannot be fixed.

- I really don't have any preferences as to which software stack I use, my goal is to use Whisper on a SM8550 (SD gen 2) running Linux. From what I've understood QNN is the only way to do that?

**Comment by escorciav** [comment]
I see. Sorry that I can't help. Did you post your question in the Slack forum associated with Qualcomm AI Hub?
perhaps consider giving a shot (deploy) the provided whisper (encoder | decoder) and make a post about it :shrug:

**Comment by quic-rneti** [comment]
Re-opening issue to address latest comments. Do consider raising a new issue.  

1. Does TensorFlow Lite Hexagon delegate works on Linux?
[Yes] 
3. Will TensorFlow Lite Hexagon delegate support the new snapdragon chips (gen 2 / 3)?
[Yes] 
5. Will you release a QNN version handling variable input lengths? So we can run .dlc transformers without using tflite?
[Not in short term plan - We understand the requirement but there is no short term ETA for this]
